google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkServiceManager.java

```diff
@@ -43,6 +77,214 @@ public void onNewCloudSdkServiceTypeSelected(CloudSdkServiceType newServiceType)
+      @Nullable Project project,
+      Runnable afterWaitComplete,
+      String progressMessage,
+      CloudSdkStatusHandler sdkLogging) {
+    CloudSdkService cloudSdkService = CloudSdkService.getInstance();
+    boolean installInProgress = isInstallInProgress();
+
+    // no need to wait for install if unsupported or completed.
+    CountDownLatch installationCompletionLatch = new CountDownLatch(installInProgress ? 1 : 0);
+
+    // listener for SDK updates, waits until install / update is done. uses latch to notify UI
+    // blocking thread.
+    final CloudSdkService.SdkStatusUpdateListener sdkStatusUpdateListener =
+        (sdkService, status) -> {
+          switch (status) {
+            case READY:
+            case INVALID:
+            case NOT_AVAILABLE:
+              installationCompletionLatch.countDown();
+              break;
+            case INSTALLING:
+              // continue waiting for completion.
+              break;
+          }
+        };
+
+    if (installInProgress) {
+      cloudSdkService.addStatusUpdateListener(sdkStatusUpdateListener);
+      sdkLogging.log(GctBundle.getString("managedsdk.waiting.for.sdk.ready") + "\n");
+
+      openBackgroundProcessWindow(project);
+    }
+
+    // wait for SDK to be ready and trigger the actual deployment if it properly installs.
+    ProgressManager.getInstance()
+        .run(
+            new Task.Backgroundable(project, progressMessage, true) {
+              @Override
+              public void run(@NotNull ProgressIndicator indicator) {
+                try {
+                  while (installationCompletionLatch.getCount() > 0) {
+                    // wait interruptibility to check for user cancel each second.
+                    installationCompletionLatch.await(1, SECONDS);
+                    if (checkIfCancelled()) {
+                      sdkLogging.onUserCancel();
+                      break;
+                    }
+                  }
+
+                } catch (InterruptedException e) {
+                  /* valid cancellation exception, no handling needed. */
+                } finally {
+                  // remove the notification listener regardless of waiting outcome.
+                  cloudSdkService.removeStatusUpdateListener(sdkStatusUpdateListener);
+                  // process logging and error notifications.
+                  ApplicationManager.getApplication().invokeLater(() -> handleErrors(sdkLogging));
+                  // run the activity after wait is over, regardless of outcome.
+                  afterWaitComplete.run();
+                }
+              }
+            });
+  }
+
+  @VisibleForTesting
+  boolean checkIfCancelled() {
+    return ProgressManager.getInstance().getProgressIndicator().isCanceled();
+  }
+
+  /**
+   * Checks if current SDK service is ready for operations. If it's not, install is attempted if
+   * supported, and true is returned.
+   */
+  private boolean isInstallInProgress() {
```

Comment:
just noticed this. I think an `isInstallInProgress` method should only return if the install is running. The side-effect of kicking off the install should probably be refactored out.

Reply:
I think it makes it more confusing to extract anything from this code block since it after calling `install()` you can't be sure exactly if SDK status has changed yet (this is async operation). I moved it back as a block, it could possibly be more convoluted if I break this 5 lines of code into two helper boolean methods and another logic using them.